File: solutions/kotlin/kindergarten-garden/1/src/main/kotlin/KindergartenGarden.kt
class KindergartenGarden(private val diagram: String) {


    private val students = listOf(
        "Alice", "Bob", "Charlie", "David",
        "Eve", "Fred", "Ginny", "Harriet",
        "Ileana", "Joseph", "Kincaid", "Larry"
    )


    private val plantMapping = mapOf(
        'G' to "grass",
        'C' to "clover",
        'R' to "radishes",
        'V' to "violets"
    )

    fun getPlantsOfStudent(student: String): List<String> {
  
        val studentIndex = students.indexOf(student)
        if (studentIndex == -1) return emptyList()


        val rows = diagram.lines()
        val row1 = rows.getOrNull(0) ?: ""
        val row2 = rows.getOrNull(1) ?: ""

    
        val firstCupIndex = studentIndex * 2
        val secondCupIndex = firstCupIndex + 1

        val result = mutableListOf<String>()

      
        if (firstCupIndex < row1.length) {
            plantMapping[row1[firstCupIndex]]?.let { result.add(it) }
        }
        if (secondCupIndex < row1.length) {
            plantMapping[row1[secondCupIndex]]?.let { result.add(it) }
        }

   
        if (firstCupIndex < row2.length) {
            plantMapping[row2[firstCupIndex]]?.let { result.add(it) }
        }
        if (secondCupIndex < row2.length) {
            plantMapping[row2[secondCupIndex]]?.let { result.add(it) }
        }

        return result
    }
}

File: solutions/typescript/kindergarten-garden/1/kindergarten-garden.ts
const DEFAULT_STUDENTS: Student[] = [
  'Alice',
  'Bob',
  'Charlie',
  'David',
  'Eve',
  'Fred',
  'Ginny',
  'Harriet',
  'Ileana',
  'Joseph',
  'Kincaid',
  'Larry',
]

const PLANT_CODES = {
  G: 'grass',
  V: 'violets',
  R: 'radishes',
  C: 'clover',
} as const

type Student = string
type Plant = (typeof PLANT_CODES)[keyof typeof PLANT_CODES]
type Plants = Plant[]

export class Garden {
  private row1: string;
  private row2: string;
  private students: Student[];

  constructor(diagram: string, students = DEFAULT_STUDENTS) {
    // Разделяем строку на два ряда растений
    const [row1, row2] = diagram.split('\n');
    this.row1 = row1;
    this.row2 = row2;
    
    // Сортируем список студентов по алфавиту, так как грядки выдаются строго по порядку
    this.students = [...students].sort();
  }

  public plants(student: Student): Plants {
    // Находим порядковый номер ребёнка в отсортированном списке
    const studentIndex = this.students.indexOf(student);
    
    // Если студент не найден, возвращаем пустой массив
    if (studentIndex === -1) {
      return [];
    }

    // Вычисляем начальный индекс символов в строке ряда для этого ребёнка
    const startIdx = studentIndex * 2;

    // Берем по 2 знака из каждого ряда
    const codes = [
      this.row1[startIdx],
      this.row1[startIdx + 1],
      this.row2[startIdx],
      this.row2[startIdx + 1],
    ] as const;

    // Переводим буквы-коды в полное название растений
    return codes.map(code => PLANT_CODES[code as keyof typeof PLANT_CODES]);
  }
}
